[CI/build] Moves source compilation to build stage

bbartels opened the pull request:
## Purpose

This PR removes the need to build NVSHMEM from source which should build times of the Docker container by around 3-5 minutes
Additionally this moves any other source builds DEEPGEMM etc from the final runtime container to the build container and copies over the resulting artifacts to the final runtime image. This allows for removing any build level tools (compilers, git, etc) from the actual final image to reduce its size.
This additionally sets the stage for any future move to flashinfer AOT mode, such that the final base image can be moved away from the `devel` to the `base` image to save space.

## Test Plan

Ensure current CI builds and tests correctly

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

bbartels committed:
Moves source compilation to build stage

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Moves flashinfer

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Merge branch 'main' into source-comp

bbartels committed:
Adds dockerfile dependency graph

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Merge branch 'main' into source-comp

cjackal commented on docker/Dockerfile:
```diff
@@ -316,7 +385,7 @@ RUN echo 'tzdata tzdata/Areas select America' | debconf-set-selections \
     fi \
     && apt-get update -y \
-    && apt-get install -y python${PYTHON_VERSION} python${PYTHON_VERSION}-dev python${PYTHON_VERSION}-venv libibverbs-dev \
+    && apt-get install -y python${PYTHON_VERSION} python${PYTHON_VERSION}-dev python${PYTHON_VERSION}-venv \
```
I think the situation is similar as in #24874. Pre-built UCX in nixl depends on libibverbs.so.1 (which I think is natural as UCX is all about communications), so it may raise runtime error in P/D disaggregation examples.

```bash
@cjackal ➜ /workspaces/vllm (fix-docs) $ ldd nixl.libs/ucx/libuct_ib.so.0
        linux-vdso.so.1 (0x00007ffd91ff8000)
        libibverbs.so.1 => not found
        libuct-2d55eade.so.0.0.0 => /workspaces/vllm/nixl.libs/ucx/../libuct-2d55eade.so.0.0.0 (0x00007ed5a3404000)
        libucs-311e600f.so.0.0.0 => /workspaces/vllm/nixl.libs/ucx/../libucs-311e600f.so.0.0.0 (0x00007ed5a3379000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ed5a3290000)
        libucm-e091ff91.so.0.0.0 => /workspaces/vllm/nixl.libs/ucx/../libucm-e091ff91.so.0.0.0 (0x00007ed5a326e000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ed5a3267000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ed5a3262000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007ed5a325d000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ed5a304b000)
```

bbartels replied:
I see, makes sense. I'll end up adding it back. Trying to figure out why my change borked the SCCACHE on CI. Doesn't seem to be working as build takes forever. Any ideas? Must be a small typo somewhere

bbartels replied:
Added back `libibverbs-dev`

bbartels committed:
Merge branch 'main' into source-comp

bbartels committed:
Fixes setup_sccache invocation

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Fixes not including libibverbs-dev in installation image

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Addresses PR feedback

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Fixes syntax

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Fixes bash script

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Fixes

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels committed:
Fixes

Signed-off-by: bbartels <benjamin@bartels.dev>

bbartels commented:
@hmellor Could I get a review on this? Build seems to pass now

bbartels commented on docker/Dockerfile:
```diff
@@ -175,7 +175,16 @@ RUN --mount=type=cache,target=/root/.cache/uv \
     --extra-index-url ${PYTORCH_CUDA_INDEX_BASE_URL}/cu$(echo $CUDA_VERSION | cut -d. -f1,2 | tr -d '.')
 
-COPY . .
+COPY setup.py setup.py
+COPY pyproject.toml pyproject.toml
+COPY cmake/ cmake/
+COPY csrc/ csrc/
+COPY requirements/ requirements
+COPY vllm/ vllm/
+COPY CMakeLists.txt CMakeLists.txt
+
+COPY tools/check_repo.sh tools/check_repo.sh
```
This helps in local builds not to invalidate cached layers. i.e. not having to rebuild vllm because you changed the script to install gdrcopy

dougbtv replied:
I also like having the whole source tree copied in too.

NickLucche reviewed (commented):
Tagging people more familiar with the current build process @mgoin @dtrifiro @wseaton

bbartels committed:
Apply suggestions from code review

Signed-off-by: Benjamin Bartels <benjamin@bartels.dev>

bbartels commented:
@dougbtv @dtrifiro @mgoin I've addressed PR feedback, could someone have another look? :) Would unblock flashinfer AOT work and moving to a slimmer base image

dtrifiro commented on docker/Dockerfile:
```diff
@@ -247,6 +242,74 @@ RUN if [ "$RUN_WHEEL_CHECK" = "true" ]; then \
     else \
         echo "Skipping wheel size check."; \
     fi
+
```
Should we also move this flashinfer build to a standalone script?

cjackal commented on tools/ep_kernels/install_python_libraries.sh:
```diff
@@ -106,19 +70,44 @@ clone_repo() {
+    popd
+}
+
+# build pplx-kernels
+do_build \
+    "https://github.com/ppl-ai/pplx-kernels" \
+    "pplx-kernels" \
+    "setup.py" \
+    "12cecfd" \
+    ""
+
+# build DeepEP
+do_build \
+    "https://github.com/deepseek-ai/DeepEP" \
+    "DeepEP" \
+    "setup.py" \
+    "abba6ad" \
+    "export NVSHMEM_DIR=$WORKSPACE/nvshmem; "
```
While not that sensitive the commits are in general, I wonder whether the dependency version bump here is intentional. And if possible I'd like to bump DeepEP version to latest for the sake of EP24, c.f. deepseek-ai/DeepEP#432.

mgoin reviewed (changes requested):
Please do not move flashinfer right now. I am trying to fix the AOT build for the release asap in https://github.com/vllm-project/vllm/pull/25730

I'm also not sure if it will work to move the flashinfer section like this since currently the `python3 -m flashinfer --download-cubin` command is just downloading files to `~/.cache/flashinfer/`

cjackal commented:
> Please do not move flashinfer right now. I am trying to fix the AOT build for the release asap in #25730
> 
> I'm also not sure if it will work to move the flashinfer section like this since currently the `python3 -m flashinfer --download-cubin` command is just downloading files to `~/.cache/flashinfer/`

Flashinfer is moving the cubin files to a separate package(`flashinfer-cubin`) and publishing it to PyPI (flashinfer-ai/flashinfer#1737), so `python3 -m flashinfer --download-cubin` will not be the preferred command since the next flashinfer major release anyway. I expect the change from the vllm side is as simple as adding `flashinfer-cubin` to the requirements.txt.
